File: cpp/src/storage/schema.cpp
// Copyright 2023 Zilliz
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "milvus-storage/storage/schema.h"
#include <memory>
#include "milvus-storage/common/constants.h"
#include "milvus-storage/common/macro.h"
#include "milvus-storage/common/utils.h"
#include "milvus-storage/common/log.h"
namespace milvus_storage {

Schema::Schema(std::shared_ptr<arrow::Schema> schema, SchemaOptions& options)
    : schema_(std::move(schema)), options_(options) {}

Status Schema::Validate() {
  RETURN_NOT_OK(options_.Validate(schema_.get()));
  RETURN_NOT_OK(BuildScalarSchema());
  RETURN_NOT_OK(BuildVectorSchema());
  RETURN_NOT_OK(BuildDeleteSchema());
  LOG_STORAGE_DEBUG_ << "Schema validate success";
  return Status::OK();
}

std::shared_ptr<arrow::Schema> Schema::schema() const { return schema_; }

const SchemaOptions& Schema::options() const { return options_; }

std::shared_ptr<arrow::Schema> Schema::scalar_schema() { return scalar_schema_; }

std::shared_ptr<arrow::Schema> Schema::vector_schema() { return vector_schema_; }

std::shared_ptr<arrow::Schema> Schema::delete_schema() { return delete_schema_; }

Result<std::unique_ptr<schema_proto::Schema>> Schema::ToProtobuf() {
  auto schema = std::make_unique<schema_proto::Schema>();
  ASSIGN_OR_RETURN_NOT_OK(auto arrow_schema, ToProtobufSchema(schema_.get()));

  auto options = options_.ToProtobuf();
  schema->set_allocated_arrow_schema(arrow_schema.release());
  schema->set_allocated_schema_options(options.release());
  return schema;
}

Status Schema::FromProtobuf(const schema_proto::Schema& schema) {
  ASSIGN_OR_RETURN_NOT_OK(schema_, FromProtobufSchema(schema.arrow_schema()));
  options_.FromProtobuf(schema.schema_options());
  RETURN_NOT_OK(BuildScalarSchema());
  RETURN_NOT_OK(BuildVectorSchema());
  RETURN_NOT_OK(BuildDeleteSchema());
  return Status::OK();
}

Status Schema::BuildScalarSchema() {
  arrow::SchemaBuilder scalar_schema_builder;
  for (const auto& field : schema_->fields()) {
    if (field->name() == options_.vector_column) {
      continue;
    }
    RETURN_ARROW_NOT_OK(scalar_schema_builder.AddField(field));
  }
  auto offset_field = std::make_shared<arrow::Field>(kOffsetFieldName, arrow::int64());
  RETURN_ARROW_NOT_OK(scalar_schema_builder.AddField(offset_field));
  ASSIGN_OR_RETURN_ARROW_NOT_OK(scalar_schema_, scalar_schema_builder.Finish());
  return Status::OK();
}

Status Schema::BuildVectorSchema() {
  arrow::SchemaBuilder vector_schema_builder;
  for (const auto& field : schema_->fields()) {
    if (field->name() == options_.primary_column || field->name() == options_.version_column ||
        field->name() == options_.vector_column) {
      RETURN_ARROW_NOT_OK(vector_schema_builder.AddField(field));
    }
  }
  ASSIGN_OR_RETURN_ARROW_NOT_OK(vector_schema_, vector_schema_builder.Finish());
  return Status::OK();
}

Status Schema::BuildDeleteSchema() {
  arrow::SchemaBuilder delete_schema_builder;
  auto pk_field = schema_->GetFieldByName(options_.primary_column);
  auto version_field = schema_->GetFieldByName(options_.version_column);
  RETURN_ARROW_NOT_OK(delete_schema_builder.AddField(pk_field));
  if (options_.has_version_column()) {
    RETURN_ARROW_NOT_OK(delete_schema_builder.AddField(version_field));
  }
  ASSIGN_OR_RETURN_ARROW_NOT_OK(delete_schema_, delete_schema_builder.Finish());
  return Status::OK();
}
}  // namespace milvus_storage

File: go/storage/schema/schema.go
package schema

import (
	"github.com/apache/arrow/go/v12/arrow"
	"github.com/milvus-io/milvus-storage-format/common"
	"github.com/milvus-io/milvus-storage-format/common/result"
	"github.com/milvus-io/milvus-storage-format/common/status"
	"github.com/milvus-io/milvus-storage-format/common/utils"
	"github.com/milvus-io/milvus-storage-format/proto/schema_proto"
	"github.com/milvus-io/milvus-storage-format/storage/options"
)

// Schema is a wrapper of arrow schema
type Schema struct {
	schema       *arrow.Schema
	scalarSchema *arrow.Schema
	vectorSchema *arrow.Schema
	deleteSchema *arrow.Schema

	options *options.SchemaOptions
}

func (s *Schema) Schema() *arrow.Schema {
	return s.schema
}

func (s *Schema) Options() *options.SchemaOptions {
	return s.options
}

func NewSchema(schema *arrow.Schema, options *options.SchemaOptions) *Schema {
	return &Schema{
		schema:  schema,
		options: options,
	}
}

func (s *Schema) Validate() status.Status {
	validate := s.options.Validate(s.schema)
	if !validate.IsOK() {
		return status.InternalStateError(validate.Msg())
	}
	scalarSchema := s.BuildScalarSchema()
	if !scalarSchema.IsOK() {
		return status.InternalStateError(scalarSchema.Msg())
	}
	vectorSchema := s.BuildVectorSchema()
	if !vectorSchema.IsOK() {
		return status.InternalStateError(vectorSchema.Msg())
	}
	deleteSchema := s.BuildDeleteSchema()
	if !deleteSchema.IsOK() {
		return status.InternalStateError(deleteSchema.Msg())
	}
	return status.OK()
}

func (s *Schema) ScalarSchema() *arrow.Schema {
	return s.scalarSchema
}

func (s *Schema) VectorSchema() *arrow.Schema {
	return s.vectorSchema
}

func (s *Schema) DeleteSchema() *arrow.Schema {
	return s.deleteSchema
}

func (s *Schema) FromProtobuf(schema *schema_proto.Schema) status.Status {
	schemaType := utils.FromProtobufSchema(schema.ArrowSchema)
	if !schemaType.Ok() {
		return status.ArrowError("invalid schema")
	}
	s.schema = schemaType.Value()
	s.options.FromProtobuf(schema.GetSchemaOptions())
	s.BuildScalarSchema()
	s.BuildVectorSchema()
	s.BuildDeleteSchema()
	return status.OK()
}

func (s *Schema) ToProtobuf() *result.Result[*schema_proto.Schema] {
	schema := &schema_proto.Schema{}
	arrowSchema := utils.ToProtobufSchema(s.schema)
	if !arrowSchema.Ok() {
		return result.NewResultFromStatus[*schema_proto.Schema](*arrowSchema.Status())
	}
	schema.ArrowSchema = arrowSchema.Value()
	schema.SchemaOptions = s.options.ToProtobuf()
	return result.NewResult[*schema_proto.Schema](schema, status.OK())
}

func (s *Schema) BuildScalarSchema() status.Status {
	fields := make([]arrow.Field, 0, len(s.schema.Fields()))
	for _, field := range s.schema.Fields() {
		if field.Name == s.options.VectorColumn {
			continue
		}
		fields = append(fields, field)
	}
	offsetFiled := arrow.Field{Name: common.KOffsetFieldName, Type: arrow.DataType(&arrow.Int64Type{})}
	fields = append(fields, offsetFiled)
	s.scalarSchema = arrow.NewSchema(fields, nil)

	return status.OK()
}

func (s *Schema) BuildVectorSchema() status.Status {
	fields := make([]arrow.Field, 0, len(s.schema.Fields()))
	for _, field := range s.schema.Fields() {
		if field.Name == s.options.VectorColumn ||
			field.Name == s.options.PrimaryColumn ||
			field.Name == s.options.VersionColumn {
			fields = append(fields, field)
		}
	}
	s.vectorSchema = arrow.NewSchema(fields, nil)

	return status.OK()
}

func (s *Schema) BuildDeleteSchema() status.Status {
	pkColumn, b := s.schema.FieldsByName(s.options.PrimaryColumn)
	if !b {
		return status.InvalidArgument("primary column not found")
	}
	versionField, b := s.schema.FieldsByName(s.options.VersionColumn)
	if !b {
		return status.InvalidArgument("version column not found")
	}
	fields := make([]arrow.Field, 0, 2)
	fields = append(fields, pkColumn[0])
	fields = append(fields, versionField[0])
	s.deleteSchema = arrow.NewSchema(fields, nil)
	return status.OK()
}
